### PR TITLE
Communications Console: The ECSining

### PR DESCRIPTION
**Pull request by Veritius**
## About the PR 
Makes the communication console ECS, as well as adding access requirements. Also added localisation support, and options for admemes.
Adds the Syndicate Communications Computer, which cannot call the shuttle.
Fixes an issue with AlertLevelSystem that caused the delay between alert changes to scale with tick rate.
Improved multi-station announcement support for a load of stuff, including salvage and certain station events.
Also moved announcement dispatches to ChatSystem.

Few things that are worth noting here.
- Anyone not standing on one of the station's grids will not receive announcements. This makes sense for being in space... but if you're floating over a hole, you also won't get it. Maybe a distance check or something? Dunno.
- There's not really a way to repair emagged comms consoles, and you can't get more, so until that's changed Command has to live with it.
- Admins can't select what station they want to announce on, it just screams out to all of them ATM. I don't really wanna touch UI so...

**Screenshots**
image

**Changelog**

:cl:
- add: Added the Syndicate Communication Console.
- tweak: Using communication consoles now requires Command access.
- tweak: Announcements are now per-station, so nukies will no longer know the state of the station they're entering.
- fix: Fix delay between alert status changes scaling with tick rate.

### CODE REVIEW

**Comment by Emisse** [comment]
can you put that into the infiltrator when you get the chance

**Comment by metalgearsloth** [comment]
You need to make the station announcement audio not broadcast if you want to support per-station consoles.

**Comment by Veritius** [comment]
I'll do that tomorrow
don't merge until then

**Comment by Veritius** [comment]
FUCK THERE'S SO MANY THINGS THAT DON'T SUPPORT MULTI STATION

**Comment by Veritius** [comment]
FUCK THERE'S SO MANY THINGS THAT NOW SUPPORT MULTI STATION

File: Content.Server/Salvage/SalvageSystem.cs [hunk 1]
```diff
@@ -58,7 +61,10 @@ private void OnGridRemoval(GridRemovalEvent ev)
             // If we ever want to give magnets names, and announce them individually, we would need to loop this, before removing it.
             if (_salvageGridStates.Remove(ev.GridId))
             {
-                Report("salvage-system-announcement-spawn-magnet-lost");
+                var gridUid = _mapManager.GetGridEuid(ev.GridId);
+                EntityManager.TryGetComponent<SalvageGridComponent>(gridUid, out var salvComp);
+                if (salvComp.SpawnerMagnet != null)
```
Comment (metalgearsloth) [inline Content.Server/Salvage/SalvageSystem.cs:66]: Just wrap this into the if statement.

File: Content.Server/StationEvents/Events/DiseaseOutbreak.cs [hunk 2]
```diff
@@ -42,6 +44,7 @@ public sealed class DiseaseOutbreak : StationEvent
     public override void Startup()
     {
         base.Startup();
+        List<EntityUid> alreadyNotifiedStations = new();
```
Comment (metalgearsloth) [inline Content.Server/StationEvents/Events/DiseaseOutbreak.cs:47]: Just use a hashset?

File: Content.Server/StationEvents/Events/RandomSentience.cs [hunk 3]
```diff
@@ -49,13 +51,27 @@ public override void Startup()
...
-                ("strength", Loc.GetString($"random-sentience-event-strength-{_random.Next(1, 8)}"))),
-                playDefaultSound: false,
-                colorOverride: Color.Gold
-        );
+
+        var _stationSystem = EntitySystem.Get<StationSystem>();
+        foreach (var target in targetList)
+        {
+            var victimOwningStation = _stationSystem.GetOwningStation(target.Owner);
+            if (victimOwningStation != null)
```
Comment (metalgearsloth) [inline Content.Server/StationEvents/Events/RandomSentience.cs:59]: You might be better off just making these Try statements with the station as an out param seeing how many there are. Also hashset again.

File: Resources/Prototypes/Entities/Mobs/Player/admin_ghost.yml [hunk 4]
```diff
@@ -97,6 +97,8 @@
           event: !type:ToggleIntrinsicUIEvent
   - type: SolarControlConsole # look ma i AM the computer!
   - type: CommunicationsConsole
+    title: communicationsconsole-announcement-title-centcom
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Mobs/Player/admin_ghost.yml:100]: snake case the communicationsconsole to comms-console

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/AlertLevel/AlertLevelComponent.cs [hunk 5]
```diff
@@ -22,12 +22,11 @@ public sealed class AlertLevelComponent : Component
...
 
+    [ViewVariables(VVAccess.ReadWrite)] public bool IsLevelLocked = false;
+    [ViewVariables] public const float Delay = 30;
     /// <summary>
     /// Is current station level can be changed by crew.
     /// </summary>
-    [ViewVariables(VVAccess.ReadWrite)] public bool IsLevelLocked = false;
-
-    [ViewVariables] public const float Delay = 300;
     [ViewVariables] public float CurrentDelay = 0;
```
Comment (metalgearsloth) [inline Content.Server/AlertLevel/AlertLevelComponent.cs:30]: These fields are cooked, you shifted the summary around.

File: Content.Server/Chat/Managers/ChatManager.cs [hunk 6]
```diff
@@ -79,16 +88,63 @@ public void DispatchServerAnnouncement(string message, Color? colorOverride = nu
...
+                }
+            }
+            else
+            {
+                var filter = Filter.Pvs(source);
+                ChatMessageToManyFiltered(filter, ChatChannel.Radio, message, messageWrap, source, true, colorOverride);
+                if (playDefaultSound)
+                {
+                    SoundSystem.Play(filter, "/Audio/Announcements/announce.ogg", AudioParams.Default.WithVolume(-2f));
+                }
```
Comment (metalgearsloth) [inline Content.Server/Chat/Managers/ChatManager.cs:144]: Should these even be on the manager and not the system? They're all sim dependent, they even add the IEntityManager dependency.

File: Content.Server/Chat/Managers/ChatManager.cs [hunk 7]
```diff
@@ -235,6 +291,21 @@ public void ChatMessageToMany(ChatChannel channel, string message, string messag
...
+            if(!filter.Recipients.Any()) return;
+
+            var clients = new List<INetChannel>();
+            foreach (var recipient in filter.Recipients)
+            {
+                clients.Add(recipient.ConnectedClient);
+            }
+            var commonSessionEnum = filter.Recipients.GetEnumerator();
+
+            ChatMessageToMany(channel, message, messageWrap, source, hideChat, clients);
```
Comment (metalgearsloth) [inline Content.Server/Chat/Managers/ChatManager.cs:306]: commonSessionEnum is unused, you can also pre-initialise the clients list size.

File: Content.Server/Communications/CommunicationsConsoleSystem.cs [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+                    if (comp.AlreadyRefreshed) continue;
+                    UpdateBoundUserInterface(comp);
+                    comp.AlreadyRefreshed = true;
+                    continue;
+                }
+                comp.AnnouncementCooldownRemaining -= frameTime;
+            }
+
+            base.Update(frameTime);
+        }
```
Comment (metalgearsloth) [inline Content.Server/Communications/CommunicationsConsoleSystem.cs:58]: Shouldn't you just check AlreadyRefreshed first? Then you don't need to touch any accumulation.

File: Content.Server/Communications/CommunicationsConsoleSystem.cs [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+        {
+            foreach (var comp in EntityManager.EntityQuery<CommunicationsConsoleComponent>())
+            {
+                var entStation = _stationSystem.GetOwningStation(comp.Owner);
+                if (args.Station == entStation)
+                {
+                    UpdateBoundUserInterface(comp);
+                }
+            }
+        }
```
Comment (metalgearsloth) [inline Content.Server/Communications/CommunicationsConsoleSystem.cs:78]: Comments

File: Content.Server/Communications/CommunicationsConsoleSystem.cs [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        {
+            return comp.CanCallShuttle && _roundEndSystem.CanCall();
+        }
+
+        private void OnSelectAlertLevelMessage(EntityUid uid, CommunicationsConsoleComponent comp, CommunicationsConsoleSelectAlertLevelMessage message)
+        {
+            if (message.Session.AttachedEntity is not {Valid: true} mob) return;
+            if (!CanUse(mob, uid))
+            {
+                _popupSystem.PopupEntity(Loc.GetString("comms-console-permission-denied"), uid, Filter.Entities(mob));
```
Comment (metalgearsloth) [inline Content.Server/Communications/CommunicationsConsoleSystem.cs:147]: Should this popup cursor? The UI might be blocking their model.

File: Content.Server/StationEvents/Events/DiseaseOutbreak.cs [hunk 11]
```diff
@@ -61,15 +64,24 @@ public override void Startup()
             return;
 
         var diseaseSystem = EntitySystem.Get<DiseaseSystem>();
-        /// Now we give it to people in the list of living disease carriers earlier
+        // Now we give it to people in the list of living disease carriers earlier
+        var _stationSystem = EntitySystem.Get<StationSystem>();
```
Comment (metalgearsloth) [inline Content.Server/StationEvents/Events/DiseaseOutbreak.cs:68]: style guide, also you're resolving IEntitySystemManager twice when you could just resolve it once first for both systems.

File: Content.Server/StationEvents/Events/DiseaseOutbreak.cs [hunk 12]
```diff
@@ -61,15 +64,24 @@ public override void Startup()
...
         foreach (var target in aliveList)
         {
             if (toInfect-- == 0)
                 break;
 
             diseaseSystem.TryAddDisease(target.Owner, disease, target);
+
+            var station = _stationSystem.GetOwningStation(target.Owner);
+            if(station == null) continue;
+            stationsToNotify.Add((EntityUid) station);
```
Comment (metalgearsloth) [inline Content.Server/StationEvents/Events/DiseaseOutbreak.cs:78]: wut

File: Content.Server/StationEvents/Events/RandomSentience.cs [hunk 13]
```diff
@@ -49,13 +51,25 @@ public override void Startup()
         var kind1 = groupList.Count > 0 ? groupList[0] : "???";
         var kind2 = groupList.Count > 1 ? groupList[1] : "???";
         var kind3 = groupList.Count > 2 ? groupList[2] : "???";
-        _chatManager.DispatchStationAnnouncement(
-            Loc.GetString("station-event-random-sentience-announcement",
-                ("kind1", kind1), ("kind2", kind2), ("kind3", kind3), ("amount", groupList.Count),
-                ("data", Loc.GetString($"random-sentience-event-data-{_random.Next(1, 6)}")),
-                ("strength", Loc.GetString($"random-sentience-event-strength-{_random.Next(1, 8)}"))),
+
+        var _stationSystem = EntitySystem.Get<StationSystem>();
```
Comment (metalgearsloth) [inline Content.Server/StationEvents/Events/RandomSentience.cs:55]: style guide.

File: SpaceStation14.sln.DotSettings [hunk 14]
```diff
@@ -1,4 +1,5 @@
 ﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
+	<s:Boolean x:Key="/Default/CodeEditing/SuppressNullableWarningFix/Enabled/@EntryValue">False</s:Boolean>
```
Comment (metalgearsloth) [inline SpaceStation14.sln.DotSettings:2]: Uhh what's this doin.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/StationEvents/Events/DiseaseOutbreak.cs (hunk 12 above)
Comment (Veritius) [reply Content.Server/StationEvents/Events/DiseaseOutbreak.cs:78]: it only notifies relevant stations (for multi-station) so if the player isn't on a station it doesn't do it

File: SpaceStation14.sln.DotSettings (hunk 14 above)
Comment (Veritius) [reply SpaceStation14.sln.DotSettings:2]: oop

File: Content.Server/Chat/Managers/ChatManager.cs (hunk 6 above)
Comment (Veritius) [reply Content.Server/Chat/Managers/ChatManager.cs:144]: It was kind of like this when I started. Mirror said they forgot to move it so :shrug:

Comment (metalgearsloth) [reply Content.Server/Chat/Managers/ChatManager.cs:144]: uhhh should probably be moved then? I don't imagine it'd be too painful. ChatManager should really only be used for OOC I think.

File: Content.Server/StationEvents/Events/DiseaseOutbreak.cs (hunk 12 above)
Comment (metalgearsloth) [reply Content.Server/StationEvents/Events/DiseaseOutbreak.cs:78]: I was referring to the cast, normally just .Value is used but I guess it's okay.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Move the dispatches to the system and also I forgot to mention but use the proxy methods for entitymanager i.e. TryComp<T> instead of EntityManager.TryGetComponent(typeof(T) etc.

File: Content.Server/Announcements/AnnounceCommand.cs [hunk 15]
```diff
@@ -13,7 +13,7 @@ public sealed class AnnounceCommand : IConsoleCommand
         public string Help => $"{Command} <sender> <message> or {Command} <message> to send announcement as centcomm.";
         public void Execute(IConsoleShell shell, string argStr, string[] args)
         {
-            var chat = IoCManager.Resolve<IChatManager>();
+            var chat = IoCManager.Resolve<ChatSystem>();
```
Comment (metalgearsloth) [inline Content.Server/Announcements/AnnounceCommand.cs:16]:
```suggestion
            var chat = IoCManager.Resolve<IEntitySystemManager>().GetEntitySystem<ChatSystem>();
```

**Pull request merged**